### DISCUSSION TITLE
Fixing town names

### DISCUSSION THREAD

**Discussion by nielsmh** [OpenTTD/OpenTTD#8737] [Features]
By now it should be well known that several of the built in town name generators have issues.

Here's a TT-Forums thread about the issue: <https://www.tt-forums.net/viewtopic.php?f=29&t=88477>

Specifically the following generators have less than 5000 possible names:

> 2820 - Hungarian
1215 - Silly
1177 - Turkish
1093 - Polish
741 - German
327 - Finnish
123 - Swiss
94 - Romanian
87 - Slovak
86 - Latin-American
70 - French

Here's some previous pull requests asking for new/fixed generators for town names:
#8683 #8152 (Spanish)
#7654 (French)
#5558 (Brazilian)
#715 #584 (German)
#261 (Catalan, last built-in town name generator added in August 2006)

Here's tickets talking about the issue of certain generators running out of names and limiting number of towns generated:
#8634 #8088 #7059 #6640 #6309 #5864 #4951

Some of the above tickets already discuss possible solutions to the issue.

I think the agreement is that the solution should be based around deprecating the old generators (keeping them around, but hidden from the GUI), and add new generators as a kind of pre-loaded NewGRF based generators.

Additionally, I think it might be an advantage to move away from storing the generator+seed of town names in savegames, and only store the name string. Make the client funding a town generate the town name and submit the final string to the command, this will allow clients to use any town name generators they have available. Two (questionable) disadvantages to this, one is losing the ability to "reset" the name of a town, the other is losing the ability to fix misspelled town names "back in time".

So here's a suggested roadmap:
- Change town founding to only send/store the final name string for creation, not the seed
- Implement way of loading town name NewGRFs statically (and automatically)
- Develop/collect town name NewGRFs that cover the above mentioned languages and include them in the official distribution of the game
- Upgrade savegames to use original generator to store the generated name, and convert old limited generator settings to the well-known replacement generators
- Deprecate and hide original generators from the menu

**Comment by LC-Zorg** [comment]
I am wondering here whether the add-ons with city names should not be downloaded to a different place than today? It's quite strange that you have to select it twice to use an add-on. First, add it to the active list and then select the appropriate option, which is currently in the map generator window (good change).
Some ideas and comments:
1.Creating a dedicated folder for add-ons with city names; moving all such add-ons from the newgrf folder to it; saving next such add-ons in it; add-ons would disappear from the NewGRF add-ons list, but they would appear in the city name options list
2. Add-ons with city names would always remain in the inactive add-ons (e.g. at the bottom under the additional line), their selection on the list of active add-ons would be impossible and would end with an appropriate message, however, these add-ons would be treated as active and would always be visible on the list city ​​name options.
3. There are other specific additions that affect the appearance of the interface. For example, NewCC, which changes, inter alia, window colors. I don't know if they shouldn't be placed elsewhere either. Maybe there should be a third window for passive additions? The ones that will be active all the time. It could be an Active / Passive NewGRF file switch.
